### PR TITLE
Updated vulnerable dependencies as reported by snyk

### PR DESCRIPTION
**Pull request by cstaylor**
All tests passed locally as well as snyk reporting no remaining critical vulnerabilities

### CODE REVIEW

**Review by haavar — COMMENTED** [review]
I'm getting a build error when trying to compile the core library on this branch. I'm not getting this error on master.

[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.owasp:dependency-check-maven:7.1.1:check (default) on project java-saml-core: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
[ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
[ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
[ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
[ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file.
[ERROR] -> [Help 1]

File: core/pom.xml [hunk 1]
```diff
@@ -72,13 +72,19 @@
 		<dependency>
 			<groupId>com.azure</groupId>
 			<artifactId>azure-security-keyvault-keys</artifactId>
-			<version>4.3.4</version>
+			<version>4.4.6</version>
 			<optional>true</optional>
+			<exclusions>
+				<exclusion>
```
Comment (haavar) [inline core/pom.xml:78]: Why did you add this exclusion?

Comment (cstaylor) [reply core/pom.xml:78]: There's no upstream fix for vulnerabilities in that library, but the unit tests all passed without it.

**Comment by cstaylor** [comment]
> I'm getting a build error when trying to compile the core library on this branch. I'm not getting this error on master.
> 
> ```
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.owasp:dependency-check-maven:7.1.1:check (default) on project java-saml-core: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
> [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file.
> [ERROR] -> [Help 1]
> ```

Try doing a `mvn clean` if you haven't already.  I had the same error until I did that.

**Comment by haavar** [comment]
> > I'm getting a build error when trying to compile the core library on this branch. I'm not getting this error on master.
> > ```
> > [INFO] ------------------------------------------------------------------------
> > [ERROR] Failed to execute goal org.owasp:dependency-check-maven:7.1.1:check (default) on project java-saml-core: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
> > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file.
> > [ERROR] -> [Help 1]
> > ```
> 
> Try doing a `mvn clean` if you haven't already. I had the same error until I did that.

I tried that, but it's still failing. I'm talking about compiling the submodule by itself, not the whole multi module project. I think the issue is that 

> > I'm getting a build error when trying to compile the core library on this branch. I'm not getting this error on master.
> > ```
> > [INFO] ------------------------------------------------------------------------
> > [ERROR] Failed to execute goal org.owasp:dependency-check-maven:7.1.1:check (default) on project java-saml-core: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
> > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file.
> > [ERROR] -> [Help 1]
> > ```
> 
> Try doing a `mvn clean` if you haven't already. I had the same error until I did that.

I have tried that, but it's not helping. The issue seems to be related the updated version of org.owasp:dependency-check-maven, not looking for `.nvd-suppressions.xml` in the right place. You may want to specify the path to this file more precise.

To be clear, I'm talking about comping the core module by itself, not the whole multi module project.

File: core/pom.xml (hunk 1 above)
Comment (haavar) [reply core/pom.xml:78]: It's also compiling fine if you don't exclude this dependency. I get that it's not needed, but there seems to be a lot of transient dependencies for `azure-security-keyvault-keys` that are not used by this project. Seems arbitrary to exclude this single one.

**Comment by haavar** [comment]
The main project also fails to build with the following message: 
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-war-plugin:2.2:war (default-war) on project java-saml-tookit-jspsample: Execution default-war of goal org.apache.maven.plugins:maven-war-plugin:2.2:war failed: Unable to load the mojo 'war' in the plugin 'org.apache.maven.plugins:maven-war-plugin:2.2' due to an API incompatibility: org.codehaus.plexus.component.repository.exception.ComponentLookupException: Cannot access defaults field of Properties`

File: core/pom.xml (hunk 1 above)
Comment (cstaylor) [reply core/pom.xml:78]: > > > I'm getting a build error when trying to compile the core library on this branch. I'm not getting this error on master.
> > > ```
> > > [INFO] ------------------------------------------------------------------------
> > > [ERROR] Failed to execute goal org.owasp:dependency-check-maven:7.1.1:check (default) on project java-saml-core: One or more exceptions occurred during dependency-check analysis: One or more exceptions occurred during analysis:
> > > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > > [ERROR] 	InitializationException: Warn initializing the suppression analyzer: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file. 
> > > [ERROR] 		caused by SuppressionParseException: Failed to load .nvd-suppressions.xml, caused by Unable to read suppression file.
> > > [ERROR] -> [Help 1]
> > > ```
> > 
> > 
> > Try doing a `mvn clean` if you haven't already. I had the same error until I did that.
> 
> I have tried that, but it's not helping. The issue seems to be related the updated version of org.owasp:dependency-check-maven, not looking for `.nvd-suppressions.xml` in the right place. You may want to specify the path to this file more precise.
> 
> To be clear, I'm talking about comping the core module by itself, not the whole multi module project.

Pushed a change that pulls the suppression file directly from github instead of path shenanigans.

Comment (cstaylor) [reply core/pom.xml:78]: I removed the exclusion.

**Comment by cstaylor** [comment]
> The main project also fails to build with the following message: `[ERROR] Failed to execute goal org.apache.maven.plugins:maven-war-plugin:2.2:war (default-war) on project java-saml-tookit-jspsample: Execution default-war of goal org.apache.maven.plugins:maven-war-plugin:2.2:war failed: Unable to load the mojo 'war' in the plugin 'org.apache.maven.plugins:maven-war-plugin:2.2' due to an API incompatibility: org.codehaus.plexus.component.repository.exception.ComponentLookupException: Cannot access defaults field of Properties`
> 
> I tried multiple versions of maven.

I'm not seeing the above error:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for OneLogin java-saml Toolkit Project 2.9.1-SNAPSHOT:
[INFO]
[INFO] OneLogin java-saml Toolkit Project ................. SUCCESS [  8.104 s]
[INFO] OneLogin java-saml Toolkit Core .................... SUCCESS [ 32.266 s]
[INFO] OneLogin java-saml Toolkit ......................... SUCCESS [ 11.135 s]
[INFO] OneLogin java-saml Toolkit Samples ................. SUCCESS [  1.214 s]
[INFO] OneLogin java-saml Toolkit Sample Webapp ........... SUCCESS [  2.522 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  55.414 s
[INFO] Finished at: 2022-08-20T09:34:13+09:00
[INFO] ------------------------------------------------------------------------
```

Can you send me the exact command line you are using when you build and see the error?

**Comment by haavar** [comment]
I'm running `mvn clean verify`. Maven version 3.8.6.

**Comment by cstaylor** [comment]
> I'm running `mvn clean verify`. Maven version 3.8.6.

```
ctaylor@mbp java-saml % mvn -v
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)
Maven home: /opt/homebrew/Cellar/maven/3.8.6/libexec
Java version: 1.8.0_322, vendor: Temurin, runtime: /Library/Java/JavaVirtualMachines/temurin-8.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
ctaylor@mbp java-saml % mvn clean verify
...
...
[INFO] Analysis Started
[INFO] Finished Archive Analyzer (0 seconds)
[INFO] Finished File Name Analyzer (0 seconds)
[INFO] Finished Jar Analyzer (0 seconds)
[INFO] Finished Dependency Merging Analyzer (0 seconds)
[INFO] Finished Version Filter Analyzer (0 seconds)
[INFO] Finished Hint Analyzer (0 seconds)
[INFO] Created CPE Index (0 seconds)
[INFO] Finished CPE Analyzer (0 seconds)
[INFO] Finished False Positive Analyzer (0 seconds)
[INFO] Finished NVD CVE Analyzer (0 seconds)
[INFO] Finished Sonatype OSS Index Analyzer (0 seconds)
[INFO] Finished Vulnerability Suppression Analyzer (0 seconds)
[INFO] Finished Dependency Bundling Analyzer (0 seconds)
[INFO] Analysis Complete (1 seconds)
[INFO] Writing report to: /Users/ctaylor/Projects/personal/java-saml/samples/java-saml-tookit-jspsample/target/dependency-check-report.html
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for OneLogin java-saml Toolkit Project 2.9.1-SNAPSHOT:
[INFO]
[INFO] OneLogin java-saml Toolkit Project ................. SUCCESS [  7.749 s]
[INFO] OneLogin java-saml Toolkit Core .................... SUCCESS [ 30.648 s]
[INFO] OneLogin java-saml Toolkit ......................... SUCCESS [ 10.960 s]
[INFO] OneLogin java-saml Toolkit Samples ................. SUCCESS [  1.284 s]
[INFO] OneLogin java-saml Toolkit Sample Webapp ........... SUCCESS [  2.276 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  53.101 s
[INFO] Finished at: 2022-08-20T11:42:27+09:00
[INFO] ------------------------------------------------------------------------
ctaylor@mbp java-saml % find . | grep war
./samples/java-saml-tookit-jspsample/target/java-saml-tookit-jspsample-2.9.1-SNAPSHOT.war
```

**Comment by cstaylor** [comment]
@haavar could you do the following and share the compressed file with me?
```
mvn -X --log-file ./mvn.log clean verify
gzip -9 mvn.log
```

**Comment by haavar** [comment]
Logs attached.

I have not looked at the poms in detail, and I'm just guessing here. Are there plugins without explicitly defined versions?

[mvn.log.gz](https://github.com/onelogin/java-saml/files/9386271/mvn.log.gz)

**Comment by cstaylor** [comment]
Hi @haavar I'm about to jump on a flight, but this looks interesting:

```
at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)nCaused by: java.io.IOException: Error while instrumenting sun/security/smartcardio/SunPCSC$ProviderService.n	at org.jacoco.agent.rt.internal_3570298.core.instr.Instrumenter.instrumentError(Instrumenter.java:160)n	at org.jacoco.agent.rt.internal_3570298.core.instr.Instrumenter.instrument(Instrumenter.java:110)n	at org.jacoco.agent.rt.internal_3570298.CoverageTransformer.transform(CoverageTransformer.java:92)n	... 62 morenCaused by: java.lang.IllegalArgumentException: Unsupported class file major version 62n
```
Do you have multiple JVMs installed?

**Comment by haavar** [comment]
I do, but only one is active at the time. I wonder if jacoco needs to be updated. I'll test when I get a chance.

**Comment by cstaylor** [comment]
> I spent some time on this today. Seems like the build is very sensitive to java versions. I was only able to build under 1.8.
> 
> Can you please make the following changes:
> 
> 1. Update jacoco version in core/pom.xml and toolkit/pom.xml to 0.8.8
> 2. In samples/java-saml-tookit-jspsample/pom.xml add org.apache.maven.plugin:maven-war-plugin:3.3.2
> 
> When I did that, the build was green for all java versions that I tested with.

I've pushed the changes you recommended.

**Comment by haavar** [comment]
Looks good to me. Thanks.

**Review by haavar — APPROVED** [review]

**Comment by cstaylor** [comment]
Thanks @haavar, what's the next step on getting this merged and a build created?

**Comment by haavar** [comment]
Not sure. I don't have "maintainer" status on this project. @pitbulk?

**Comment by pitbulk** [comment]
Addressed in:
- https://github.com/SAML-Toolkits/java-saml/pull/404
- https://github.com/SAML-Toolkits/java-saml/pull/405
